state/protocol/protocol_state/mutator.go

```diff
@@ -0,0 +1,58 @@
+package protocol_state
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/transaction"
+)
+
+// Mutator implements protocol.StateMutator interface.
+// It has to be used for each block to update the protocol state, even if there are no changes incorporated in candidate block.
+// Such requirement is due to the fact that protocol state is indexed by block ID, and we need to maintain such index.
+type Mutator struct {
+	protocolStateDB storage.ProtocolState
+}
+
+var _ protocol.StateMutator = (*Mutator)(nil)
+
+func NewMutator(protocolStateDB storage.ProtocolState) *Mutator {
+	return &Mutator{
+		protocolStateDB: protocolStateDB,
+	}
+}
+
+// CreateUpdater creates a new protocol state updater for the given candidate block.
+// Has to be called for each block to correctly index the protocol state.
+// No errors are expected during normal operations.
+func (m *Mutator) CreateUpdater(candidate *flow.Header) (protocol.StateUpdater, error) {
+	parentState, err := m.protocolStateDB.ByBlockID(candidate.ParentID)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve protocol state for block (%v): %w", candidate.ParentID, err)
+	}
+	return newUpdater(candidate, parentState), nil
+}
+
+// CommitProtocolState commits the protocol state updater as part of DB transaction.
+// Has to be called for each block to correctly index the protocol state.
+// No errors are expected during normal operations.
+func (m *Mutator) CommitProtocolState(updater protocol.StateUpdater) func(tx *transaction.Tx) error {
+	return func(tx *transaction.Tx) error {
+		updatedState, updatedStateID, hasChanges := updater.Build()
+		if hasChanges {
+			err := m.protocolStateDB.StoreTx(updatedStateID, updatedState)(tx)
```

Comment:
Suppose we have two protocol state changes, $\delta_1$, $\delta_2$, and applying them sequentially to an initial state $\Delta$ results in $\Delta$ again.

$\Delta + \delta_1 = \Delta'$
$\Delta' + \delta_2 = \Delta$

For example, if the state changes subtract, then add, a constant value to `Weight`.

In that case, we will have an unhandled `storage.ErrAlreadyExists` error here, treated as irrecoverable. I don't think we should make the assumption at this layer that protocol states can never repeat, even if it happens to be true. 

**Suggestion:** Handle `ErrAlreadyExists` here, and consider as a no-op.

Reply:
+1

We should certainly document possible error returns in detail for the [interface `ProtocolState`](https://github.com/onflow/flow-go/blob/ddd1a1c485ad32e7547b604431d7d9ed696c1026/storage/protocol_state.go). See [my comment below](https://github.com/onflow/flow-go/pull/4597#issuecomment-1709526474) for a detailed suggestion

Reply:
⚠️  Approving the PR - but please address this before merging.

Reply:
but in such case `hasChanges` has to be false, so it shouldn't happen. Nevertheless I will make it accept such case. [`7266c8f` (#4597)](https://github.com/onflow/flow-go/pull/4597/commits/7266c8f0adf0de4f3ac8fdc04ff2d7180f05b585)